configure: Fix cross compiling with gen-internal-compose-table

oreo639 opened the pull request:
Use AX_PROG_CC_FOR_BUILD to get build CC/CFLAGS/LDFLAGS/etc. Use AC_COMPILE_IFELSE insead of AC_RUN_IFELSE.

https://github.com/ibus/ibus/issues/2479

fujiwarat commented:
Interesting patch.
Do you have the different values of CC_FOR_BUILD and LDFLAGS_FOR_BUILD in different between x86_64 and aarch64?

fujiwarat commented on configure.ac:
```diff
@@ -131,6 +131,10 @@ AM_PROG_VALAC([0.20])
 
+m4_ifdef([AX_PROG_CC_FOR_BUILD], [
+AX_PROG_CC_FOR_BUILD
+], [AC_FATAL([AX_PROG_CC_FOR_BUILD not found, you will need to install autoconf-archive])])
```
This makes users to install autoconf-archive forcibly. Do you have any ideas to put a warning only that the flag might be important for cross-compile instead of the fatal error? I guess `$(CC_FOR_BUILD)` could be applied to CC in case `$(CC_FOR_BUILD)` is not NULL.

oreo639 replied:
The forces the user to install autoconf-archive when doing `autoreconf`, if they are just doing `./configure` without autoconf-archive, it shouldn't display an error (and autoconf-archive shouldn't be necessary after generating `./configure`). Is that fine?

fujiwarat commented on src/Makefile.am:
```diff
@@ -196,6 +198,11 @@ gen_internal_compose_table_SOURCES = \
+$(GEN_INTERNAL_COMPOSE_TABLE): CFLAGS=$(CFLAGS_FOR_BUILD)
+$(GEN_INTERNAL_COMPOSE_TABLE): CPPFLAGS=$(CPPFLAGS_FOR_BUILD)
+$(GEN_INTERNAL_COMPOSE_TABLE): LDFLAGS=$(LDFLAGS_FOR_BUILD)
 gen_internal_compose_table_LDADD = \
```
Do you have your problem for gen-install-compose-table only?
I wonder why you don't think the same issue in emoji-parser and unicode-parser.

oreo639 replied:
Oh, you're right. We disabled emoji-parser and unicode-parser when cross-compiling.

fujiwarat replied:
Probably I think ibus cannot support the multi architectures. libibus is used in ibus/binding/vala, ibus/ui/gtk3, ibus/tools besides ibus/src but I don't know if it makes sense to support the multi architectures for gen-install-compose-table only.
I guess you could put `CC=${CC_FOR_BUILD}` in configure.ac or disable `configure --host` option?

oreo639 replied:
Maybe I can look into implementing it similar to how Mesa used to do it (in a few days):
<https://github.com/Mesa3D/mesa/blob/mesa-9.1.6/src/glsl/builtin_compiler/Makefile.am>

oreo639 replied:
Should be fixed now.

fujiwarat reviewed (changes requested):
Could you add a link of https://sourceware.org/autobook/autobook/autobook_143.html#Supporting-Cross-Compiler-in-Make in the commit description?

Also please add "BUG=https://github.com/ibus/ibus/issues/2479" in the commit description too.

oreo639 commented:
Alright, sorry for the delay.

I added `ax_prog_cc_for_build.m4` to the repo to avoid adding a dependency on autoconf-archive, let me know if that is a problem.

I also updated it so now emoji-parser and unicode-parser should cross compile correctly as well.

oreo639 commented:
We use wrapper scripts to allow gir to work properly when cross compiling:
https://github.com/void-linux/void-packages/blob/master/srcpkgs/gobject-introspection/files/g-ir-scanner-qemuwrapper

Here is the template for reference:
https://github.com/void-linux/void-packages/blob/a69748d491ca42520d865e63c0d574d1444c8fde/srcpkgs/ibus/template

fujiwarat commented:
I updated configure.ac for BSD systems. Please rebase your patch.

fujiwarat commented:
Thank you for your patch and LGTM.
Since the change is big, I will integrate this patch to Fedora internally at first and integrate it to upstream if we don't see any regressions.

I still wonder if we could fix the issue without the code changes with `if CROSS_COMPILE` but set build parameters out of the ibus codes likes [g-ir-scanner-wrapper](https://github.com/void-linux/void-packages/blob/a279abf769ef7cab6ebb49d8c99bf774a2ff97b8/srcpkgs/gobject-introspection/files/g-ir-scanner-wrapper) because this change to maintain both `if CROSS_COMPILE`  and else and it's a bit complicated to me.

I also wish to integrate the g-ir-scanner-wrapper into ibus/test for the reference.

oreo639 commented:
Merged in https://github.com/ibus/ibus/commit/18f0d9f5f245f92dee086ce57a5bb94e63f1c222

fujiwarat commented:
I had integrated the changes in the Fedora rawhide for a while and decided to integrate this patch.
Thank you for the patch.

oreo639 commented:
No problem. Sorry for the complexity, if I run into any issues I will let you know, if I become aware of a better solution, I will let you know.
